dm/syncer/data_validator.go

```diff
@@ -764,7 +836,17 @@ func getRowChangeType(t replication.EventType) rowChangeType {
 }
 
 func genRowKey(pkValues []string) string {
-	return strings.Join(pkValues, "-")
+	// TODO: in scenario below, the generated key may not unique, but it's rare
+	// suppose a table with multiple column primary key: (v1, v2)
+	// for below case, the generated key is the same:
+	// 	 (aaa\t, bbb) and (aaa, \tbbb), the joint values both are "aaa\t\tbbb"
+	join := strings.Join(pkValues, "\t")
+	// if the key is too long, need to make sure it can be stored into database
+	if len(join) > maxRowKeyLength {
+		sum := sha256.Sum256([]byte(join))
+		return hex.EncodeToString(sum[:])
```

Comment:
why we do this opt ? can we just report an error? or we can change the meta schema ?

Reply:
pk of user table is part of our pk in meta-schema, so if this pk is long enough(see 3000 bytes), our pk will exceed the limit of mysql(3072 bytes), so make it a hash digest then we can store.

in most case the pk of user table is bigint, it should not go into this branch.